### PR TITLE
os/bluestore: fsck: verify blob.unused field

### PR DESCRIPTION
**Pull request by liewegas**
Two checks:

- verify that no logical extents reference the portion of the
  blob marked unused, and
- verify that the csum (if present) is zero for any unused
  region.

And a few fixes to the blob reuse code that the checks turn up!

### CODE REVIEW

**Comment by liewegas** [comment]
@ifed01 can you take a look?

**Comment by ifed01** [comment]
@liewegas - will do

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-04-05_05:26:00-rados-wip-sage-testing---basic-smithi/

**Comment by liewegas** [comment]
@ifed01 look ok?  passes ceph_test_objectstore and rados run

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1784,6 +1784,11 @@ bool BlueStore::Blob::try_reuse_blob(uint32_t min_alloc_size,
       return false;
     }
 
+    // FIXME: in some cases we could reduce unused resolution
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Suggest to move on top of the func, close to is_mutable check...

Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Never mind, that's for extend only...

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -5206,6 +5207,65 @@ int BlueStore::fsck(bool deep)
 	  l.blob_offset, 
 	  l.length);
 	++num_extents;
+	if (blob.has_unused()) {
+	  auto p = referenced.find(l.blob);
+	  uint16_t *pu;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: make a typedef for unused type to ease subsequent evolution if any?

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -5206,6 +5207,65 @@ int BlueStore::fsck(bool deep)
...
+	}
+      }
+      for (auto &i : referenced) {
+	dout(20) << __func__ << "  referenced 0x" << std::hex << i.second
+		 << std::dec << " for " << *i.first << dendl;
+	const bluestore_blob_t& blob = i.first->get_blob();
+	if (i.second & blob.unused) {
+	  derr << __func__ << " error: " << oid << " blob claims unused 0x"
+	       << std::hex << blob.unused
+	       << " but extents referece 0x" << i.second
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: typo

**Comment by liewegas** [comment]
pushed fixes

**Comment by liewegas** [comment]
thanks!

**Pull request merged**